File: client/src/lib/games/isolation/partition.ts
/**
 * Partition Detection for ISOLATION
 *
 * Detects when the board is partitioned (players cannot reach each other)
 * and calculates the size of each player's region.
 *
 * FIXED: Now uses queen-movement based reachability instead of adjacency.
 * This correctly identifies when pieces are truly isolated.
 */

import type { BoardState } from "./types";
import {
  detectPartitionBitboard,
  queenFloodFill,
  createBlockedBitboard,
  posToIndex,
  indexToPos,
  bitboardToIndices,
  popCount,
  CELL_MASKS,
  getQueenMoves,
  BOARD_SIZE
} from "./bitboard";

// 8 directions for queen movement
const DIRECTIONS = [
  { dr: -1, dc: -1 }, { dr: -1, dc: 0 }, { dr: -1, dc: 1 },
  { dr: 0, dc: -1 },                      { dr: 0, dc: 1 },
  { dr: 1, dc: -1 },  { dr: 1, dc: 0 },  { dr: 1, dc: 1 }
];

export interface PartitionResult {
  isPartitioned: boolean;
  playerRegionSize: number;
  aiRegionSize: number;
  predictedWinner: 'player' | 'ai' | 'tie' | 'unknown';
  // Additional info for endgame solver
  playerReachableCells: Set<string>;
  aiReachableCells: Set<string>;
}

/**
 * Detect if the board is partitioned and calculate region sizes
 * Uses queen-movement based reachability for accuracy
 */
export function detectPartition(board: BoardState): PartitionResult {
  const { boardSize, playerPos, aiPos, destroyed } = board;

  // Use bitboard-based partition detection
  const result = detectPartitionBitboard(playerPos, aiPos, destroyed);

  if (!result.isPartitioned) {
    return {
      isPartitioned: false,
      playerRegionSize: 0,
      aiRegionSize: 0,
      predictedWinner: 'unknown',
      playerReachableCells: new Set(),
      aiReachableCells: new Set()
    };
  }

  // Convert bitboard regions to sets for compatibility
  const playerReachableCells = new Set<string>();
  const aiReachableCells = new Set<string>();

  const playerIndices = bitboardToIndices(result.playerRegion);
  const aiIndices = bitboardToIndices(result.aiRegion);

  for (const idx of playerIndices) {
    const pos = indexToPos(idx);
    playerReachableCells.add(`${pos.r},${pos.c}`);
  }

  for (const idx of aiIndices) {
    const pos = indexToPos(idx);
    aiReachableCells.add(`${pos.r},${pos.c}`);
  }

  // Calculate predicted winner based on region sizes
  let predictedWinner: 'player' | 'ai' | 'tie' | 'unknown';

  if (result.playerRegionSize > result.aiRegionSize) {
    predictedWinner = 'player';
  } else if (result.aiRegionSize > result.playerRegionSize) {
    predictedWinner = 'ai';
  } else {
    // Equal region sizes - more complex analysis needed
    // Generally, equal regions favor the player who moves second
    // But it depends on exact board shape
    predictedWinner = 'tie';
  }

  return {
    isPartitioned: true,
    playerRegionSize: result.playerRegionSize,
    aiRegionSize: result.aiRegionSize,
    predictedWinner,
    playerReachableCells,
    aiReachableCells
  };
}

/**
 * Legacy flood fill for compatibility (now uses queen movement)
 */
function floodFillQueen(
  startPos: { r: number; c: number },
  blockedPiecePos: { r: number; c: number },
  boardSize: { rows: number; cols: number },
  destroyed: { r: number; c: number }[]
): Set<string> {
  // Create blocked bitboard
  let blocked = 0n;
  for (const d of destroyed) {
    blocked |= CELL_MASKS[posToIndex(d.r, d.c)];
  }
  blocked |= CELL_MASKS[posToIndex(blockedPiecePos.r, blockedPiecePos.c)];

  // Use queen flood fill
  const reachable = queenFloodFill(startPos, blocked);

  // Convert to set
  const result = new Set<string>();
  const indices = bitboardToIndices(reachable);
  for (const idx of indices) {
    const pos = indexToPos(idx);
    if (pos.r !== startPos.r || pos.c !== startPos.c) {
      result.add(`${pos.r},${pos.c}`);
    }
  }

  return result;
}

/**
 * Check if a move would cause a partition
 * Useful for strategic decisions about when to isolate
 */
export function wouldCausePartition(
  board: BoardState,
  destroyPos: { r: number; c: number }
): boolean {
  const { boardSize, playerPos, aiPos, destroyed } = board;

  // Create temporary destroyed list with the new destroy position
  const tempDestroyed = [...destroyed, destroyPos];

  // Check if partition would occur
  const result = detectPartitionBitboard(playerPos, aiPos, tempDestroyed);
  return result.isPartitioned;
}

/**
 * Evaluate partition potential - how close is the board to being partitioned
 * Returns a score where higher = more likely to partition soon
 */
export function evaluatePartitionPotential(board: BoardState): number {
  const { boardSize, playerPos, aiPos, destroyed } = board;

  // Create blocked bitboard
  let blocked = 0n;
  for (const d of destroyed) {
    blocked |= CELL_MASKS[posToIndex(d.r, d.c)];
  }

  // Calculate the "choke points" between players
  // These are cells that, if destroyed, would isolate one player

  const playerIdx = posToIndex(playerPos.r, playerPos.c);
  const aiIdx = posToIndex(aiPos.r, aiPos.c);

  // Count potential partition-causing cells
  let partitionCells = 0;
  let totalChecked = 0;

  for (let r = 0; r < boardSize.rows; r++) {
    for (let c = 0; c < boardSize.cols; c++) {
      const idx = posToIndex(r, c);

      // Skip destroyed and occupied cells
      if (blocked & CELL_MASKS[idx]) continue;
      if (idx === playerIdx || idx === aiIdx) continue;

      totalChecked++;

      // Check if destroying this cell would partition
      const tempDestroyed = [...destroyed, { r, c }];
      const result = detectPartitionBitboard(playerPos, aiPos, tempDestroyed);
      if (result.isPartitioned) {
        partitionCells++;
      }
    }
  }

  if (totalChecked === 0) return 1.0;

  // Return ratio of partition-causing cells
  return partitionCells / totalChecked;
}

/**
 * Find the best cell to destroy to maximize partition advantage
 * Returns null if no advantageous partition is possible
 */
export function findBestPartitionDestroy(
  board: BoardState,
  isAI: boolean
): { r: number; c: number } | null {
  const { boardSize, playerPos, aiPos, destroyed } = board;

  let bestDestroy: { r: number; c: number } | null = null;
  let bestAdvantage = -Infinity;

  const playerIdx = posToIndex(playerPos.r, playerPos.c);
  const aiIdx = posToIndex(aiPos.r, aiPos.c);

  // Create base blocked bitboard
  let blocked = 0n;
  for (const d of destroyed) {
    blocked |= CELL_MASKS[posToIndex(d.r, d.c)];
  }

  // Try destroying each empty cell
  for (let r = 0; r < boardSize.rows; r++) {
    for (let c = 0; c < boardSize.cols; c++) {
      const idx = posToIndex(r, c);

      // Skip if already destroyed or occupied
      if (blocked & CELL_MASKS[idx]) continue;
      if (idx === playerIdx || idx === aiIdx) continue;

      const destroyPos = { r, c };

      // Check if this would cause partition
      const tempDestroyed = [...destroyed, destroyPos];
      const result = detectPartitionBitboard(playerPos, aiPos, tempDestroyed);

      if (result.isPartitioned) {
        // Calculate the advantage (positive = good for AI)
        const advantage = result.aiRegionSize - result.playerRegionSize;

        // If we're AI, we want positive advantage
        // If we're player, we want negative advantage
        const adjustedAdvantage = isAI ? advantage : -advantage;

        if (adjustedAdvantage > bestAdvantage) {
          bestAdvantage = adjustedAdvantage;
          bestDestroy = destroyPos;
        }
      }
    }
  }

  // Only return if we found an advantageous partition
  if (bestAdvantage > 0) {
    return bestDestroy;
  }

  return null;
}

/**
 * Calculate the critical cells - cells that would severely impact territory
 * These are the "choke points" of the board
 */
export function findCriticalCells(board: BoardState): { r: number; c: number }[] {
  const { boardSize, playerPos, aiPos, destroyed } = board;
  const critical: { r: number; c: number }[] = [];

  const playerIdx = posToIndex(playerPos.r, playerPos.c);
  const aiIdx = posToIndex(aiPos.r, aiPos.c);

  let blocked = 0n;
  for (const d of destroyed) {
    blocked |= CELL_MASKS[posToIndex(d.r, d.c)];
  }

  for (let r = 0; r < boardSize.rows; r++) {
    for (let c = 0; c < boardSize.cols; c++) {
      const idx = posToIndex(r, c);
      if (blocked & CELL_MASKS[idx]) continue;
      if (idx === playerIdx || idx === aiIdx) continue;

      // Check if destroying this would cause partition or significantly reduce territory
      const tempDestroyed = [...destroyed, { r, c }];
      const result = detectPartitionBitboard(playerPos, aiPos, tempDestroyed);

      if (result.isPartitioned) {
        critical.push({ r, c });
      }
    }
  }

  return critical;
}

File: client/src/lib/games/isolation/wasm/src/partition.rs
//! Partition Detection for ISOLATION
//!
//! Detects when the board is partitioned (players cannot reach each other)
//! and calculates the size of each player's region.
//!
//! This is CRITICAL for strategic evaluation (weight 500).
//!
//! Based on TypeScript implementation in bitboard.ts:detectPartitionBitboard()

use crate::board::GameState;
use crate::bitboard::*;

#[derive(Clone, Copy, Debug)]
pub struct PartitionResult {
    pub is_partitioned: bool,
    pub player_region_size: i32,
    pub ai_region_size: i32,
    pub player_region: u64,
    pub ai_region: u64,
}

/// Detect if the board is partitioned using queen-based flood fill
///
/// A partition occurs when:
/// - Player cannot reach AI position via any sequence of queen moves
/// - AI cannot reach player position via any sequence of queen moves
///
/// Returns partition status and region sizes for both players
pub fn detect_partition_bitboard(
    player_pos: (u8, u8),
    ai_pos: (u8, u8),
    destroyed: u64,
) -> PartitionResult {
    let player_idx = pos_to_index(player_pos.0, player_pos.1);
    let ai_idx = pos_to_index(ai_pos.0, ai_pos.1);

    let player_mask = 1u64 << player_idx;
    let ai_mask = 1u64 << ai_idx;

    // Flood fill from player position, treating AI position as BLOCKED
    // This gives us all cells the player can reach
    let blocked_for_player = destroyed | ai_mask;
    let player_reachable = queen_flood_fill(player_pos, blocked_for_player);

    // Check if AI position is in player's reachable set
    let is_partitioned = (player_reachable & ai_mask) == 0;

    if !is_partitioned {
        // Players can reach each other - no partition
        return PartitionResult {
            is_partitioned: false,
            player_region_size: 0,
            ai_region_size: 0,
            player_region: 0,
            ai_region: 0,
        };
    }

    // Board IS partitioned - calculate region sizes
    // Player's region = all cells reachable from player (includes player position)
    let player_region = player_reachable | player_mask;

    // AI's region = all cells reachable from AI (includes AI position)
    let blocked_for_ai = destroyed | player_mask;
    let ai_reachable = queen_flood_fill(ai_pos, blocked_for_ai);
    let ai_region = ai_reachable | ai_mask;

    PartitionResult {
        is_partitioned: true,
        player_region_size: count_ones(player_region) as i32,
        ai_region_size: count_ones(ai_region) as i32,
        player_region,
        ai_region,
    }
}

/// Queen-based flood fill using bitboards
///
/// Returns all cells reachable by any sequence of queen moves from start_pos.
/// The blocked bitboard indicates cells that cannot be moved through.
pub fn queen_flood_fill(start_pos: (u8, u8), blocked: u64) -> u64 {
    let start_idx = pos_to_index(start_pos.0, start_pos.1);
    let start_mask = 1u64 << start_idx;

    let mut reachable = start_mask;
    let mut frontier = start_mask;

    let max_iterations = 50; // Safety limit
    let mut iterations = 0;

    while frontier != 0 && iterations < max_iterations {
        iterations += 1;
        let mut new_frontier = 0u64;

        // For each cell in frontier, get all queen moves
        let mut temp = frontier;
        while temp != 0 {
            let lowest_bit = temp & temp.wrapping_neg();
            let idx = lowest_bit.trailing_zeros() as u8;
            let (r, c) = index_to_pos(idx);

            // Get moves from this position
            // Block visited cells to avoid infinite expansion
            let moves = get_queen_moves(r, c, blocked | reachable);
            new_frontier |= moves & !reachable;

            temp &= temp - 1;
        }

        reachable |= new_frontier;
        frontier = new_frontier;
    }

    reachable
}

/// Check if destroying a specific cell would cause partition
///
/// Useful for strategic decisions about when to isolate opponent
pub fn would_cause_partition(
    player_pos: (u8, u8),
    ai_pos: (u8, u8),
    destroyed: u64,
    destroy_pos: (u8, u8),
) -> bool {
    let destroy_idx = pos_to_index(destroy_pos.0, destroy_pos.1);
    let destroy_mask = 1u64 << destroy_idx;
    let new_destroyed = destroyed | destroy_mask;

    let result = detect_partition_bitboard(player_pos, ai_pos, new_destroyed);
    result.is_partitioned
}

/// Evaluate partition potential - how close is the board to being partitioned
///
/// Checks critical cells and returns a score (0.0 to 1.0) where higher = more likely to partition soon
pub fn evaluate_partition_potential(
    player_pos: (u8, u8),
    ai_pos: (u8, u8),
    destroyed: u64,
) -> f32 {
    let full_board = (1u64 << CELL_COUNT) - 1;
    let player_idx = pos_to_index(player_pos.0, player_pos.1);
    let ai_idx = pos_to_index(ai_pos.0, ai_pos.1);
    let player_mask = 1u64 << player_idx;
    let ai_mask = 1u64 << ai_idx;

    let occupied = destroyed | player_mask | ai_mask;
    let empty = full_board & !occupied;

    if empty == 0 {
        return 1.0; // No empty cells, likely already partitioned
    }

    // Count cells that would cause partition if destroyed
    let mut partition_cells = 0;
    let mut total_checked = 0;

    let mut temp = empty;
    while temp != 0 {
        let lowest_bit = temp & temp.wrapping_neg();
        let idx = lowest_bit.trailing_zeros() as u8;
        let (r, c) = index_to_pos(idx);

        total_checked += 1;

        // Check if destroying this cell would partition
        if would_cause_partition(player_pos, ai_pos, destroyed, (r, c)) {
            partition_cells += 1;
        }

        temp &= temp - 1;
    }

    if total_checked == 0 {
        return 1.0;
    }

    partition_cells as f32 / total_checked as f32
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_no_partition_empty_board() {
        // Empty board - no partition
        let player_pos = (0, 0);
        let ai_pos = (6, 6);
        let destroyed = 0u64;

        let result = detect_partition_bitboard(player_pos, ai_pos, destroyed);
        assert!(!result.is_partitioned);
    }

    #[test]
    fn test_partition_diagonal_wall() {
        // Create a diagonal wall separating players
        let player_pos = (0, 0);
        let ai_pos = (6, 6);

        let mut destroyed = 0u64;
        for i in 1..6 {
            destroyed |= 1u64 << pos_to_index(i, i);
        }

        let result = detect_partition_bitboard(player_pos, ai_pos, destroyed);
        assert!(result.is_partitioned);
        assert!(result.player_region_size > 0);
        assert!(result.ai_region_size > 0);
    }

    #[test]
    fn test_queen_flood_fill() {
        // Test flood fill from center with no obstacles
        let start_pos = (3, 3);
        let blocked = 0u64;

        let reachable = queen_flood_fill(start_pos, blocked);

        // Should reach all 49 cells (including start)
        assert_eq!(count_ones(reachable), 49);
    }

    #[test]
    fn test_queen_flood_fill_blocked() {
        // Test flood fill with diagonal wall
        let start_pos = (0, 0);

        let mut blocked = 0u64;
        for i in 0..7 {
            blocked |= 1u64 << pos_to_index(i, i);
        }

        let reachable = queen_flood_fill(start_pos, blocked);

        // Should reach less than full board due to wall
        assert!(count_ones(reachable) < 49);
        assert!(count_ones(reachable) > 0);
    }

    #[test]
    fn test_would_cause_partition() {
        // Test if destroying a specific cell causes partition
        let player_pos = (0, 0);
        let ai_pos = (6, 6);

        let mut destroyed = 0u64;
        for i in 1..5 {
            destroyed |= 1u64 << pos_to_index(i, i);
        }

        // Destroying (5,5) should complete the diagonal wall and partition
        let would_partition = would_cause_partition(
            player_pos,
            ai_pos,
            destroyed,
            (5, 5),
        );
        assert!(would_partition);

        // Destroying (0,1) should NOT cause partition
        let would_not_partition = would_cause_partition(
            player_pos,
            ai_pos,
            destroyed,
            (0, 1),
        );
        assert!(!would_not_partition);
    }
}
